internal/pkg/runtime/launcher/native/launcher_linux.go

```diff
@@ -875,25 +875,29 @@ func (l *Launcher) setNamespaces() {
 
 // setEnv sets the environment for the container, from the host environment, glads, env-file.
 func (l *Launcher) setEnv(ctx context.Context, args []string) error {
-	if l.cfg.EnvFile != "" {
+	if len(l.cfg.EnvFiles) > 0 {
 		currentEnv := append(
 			os.Environ(),
 			"SINGULARITY_IMAGE="+l.engineConfig.GetImage(),
 		)
 
-		env, err := env.FileMap(ctx, l.cfg.EnvFile, args, currentEnv)
-		if err != nil {
-			return fmt.Errorf("while processing %s: %w", l.cfg.EnvFile, err)
+		// Read all environment files and put the variables into envFilesMap,
+		// environment variables in later files will take precedence.
+		envFilesMap := map[string]string{}
+		for _, envFile := range l.cfg.EnvFiles {
+			tempEnvMap, err := env.FileMap(ctx, envFile, args, currentEnv)
+			if err != nil {
+				return fmt.Errorf("while processing %s: %w", envFile, err)
+			}
+			envFilesMap = env.MergeEnvFileMap(envFilesMap, tempEnvMap, envFile)
```

Comment:
It's a little bit unusual to pass the filename `envFile` here, when it's only used for log messages.

When you see the function name `MergeEnvFileMap` and it is passed `envFile`, it would be easy to think that the function does something with the envFile itself.

I would suggest moving the debug message for envFile into the loop, and just using `MergeMap`. When multiple `envFiles` are in use, I don't think we need to warn about values being overridden between them, as we have clearly documented the precedence... and it's intuitive when they come in a particular order on the command line.

The existing message below about `--env` overriding `--env-file` is more useful... it's less obvious which order they would apply.

Reply:
I agree, I don't feel completely comfortable with that solution. I wasn't sure if the warning was really needed and just ended being conservative about the previous code. I also agree that the documentation should be enough.
I'll work on this, thanks.